### PR TITLE
feat: add/remove label in Python SDK [MLG-729]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

adds experiment label support to Python SDK with `add_label`, `remove_label`, and `set_labels`. adds new backend REST APIs to support this functionality.

## Test Plan

```
exp = client.get_experiment(ID)
```
- `exp.add_label` should add label to experiment's labels (`exp.labels`)
- `exp.remove_label` should remove label from experiment's labels (`exp.labels`)
- `exp.set_labels` should set labels on experiment's labels (`exp.labels`) and overwrite any existing labels

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/api_experiment.go [hunk 1]
```diff
@@ -2337,3 +2338,109 @@ func (a *apiServer) CreateTrial(
...
+		Labels: exp.Labels,
+	}
+	marshalledPatches, patchErr := json.Marshal(patch)
+	if patchErr != nil {
+		return nil, status.Errorf(codes.Internal, "failed to marshal experiment patch")
+	}
+
+	_, err = a.m.db.RawQuery(
+		"patch_experiment", exp.Id, marshalledPatches, exp.Notes)
+	if err != nil {
```
Comment (stoksc) [inline master/internal/api_experiment.go:2390]: running into the same problem here as the original problem, but with the span of time it can happen in from line 2350 to 2390. I would just write a new query to patch the labels, or do all this in a transaction.

File: master/internal/api_experiment.go [hunk 2]
```diff
@@ -2337,3 +2338,109 @@ func (a *apiServer) CreateTrial(
...
+	expLabels := set.New[string]()
+	for _, label := range exp.Labels {
+		expLabels.Insert(label)
+	}
+	if expLabels.Contains(req.Label) {
+		return &apiv1.PutExperimentLabelResponse{Labels: exp.Labels}, nil
+	}
+	expLabels.Insert(req.Label)
+
+	exp.Labels = expLabels.ToSlice()
```
Comment (stoksc) [inline master/internal/api_experiment.go:2375]: Equally clear, to me
```suggestion
	if slices.Contains(exp.Labels, req.Label) {
		return &apiv1.PutExperimentLabelResponse{Labels: exp.Labels}, nil
	}
	exp.Labels = append(exp.Labels, req.Label)
```

File: master/internal/api_experiment.go (hunk 1 above)
Comment (stoksc) [reply master/internal/api_experiment.go:2390]: At least before, it is semi obvious that determined would not be doing transactions over multiple API calls in the flow GET /experiments/:id, add my label, PATCH /experiments/:id. Losing updates in the course of a single API call is a less obvious/acceptable.

Comment (azhou-determined) [reply master/internal/api_experiment.go:2390]: you're totally right, i tried to follow `PatchExperiments` and blindly assumed that the transaction was somehow magically handled 🤦 will fix this.

File: harness/tests/determined/common/experimental/test_experiment.py [hunk 3]
```diff
@@ -147,3 +149,39 @@ def test_list_trials_requests_pages_lazily(
         assert len(responses.calls) == page_num
     total_pages = math.ceil((len(tr_resp.trials)) / page_size)
     assert len(responses.calls) == total_pages
+
+
+@responses.activate
+@mock.patch("determined.common.api.bindings.put_PutExperimentLabel")
+def test_add_label_url_encodes_label(
```
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:156]: I like your test.

File: harness/tests/determined/common/experimental/test_experiment.py [hunk 4]
```diff
@@ -147,3 +149,39 @@ def test_list_trials_requests_pages_lazily(
         assert len(responses.calls) == page_num
     total_pages = math.ceil((len(tr_resp.trials)) / page_size)
     assert len(responses.calls) == total_pages
+
+
+@responses.activate
```
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:154]: Don't think the decorator is needed (or helpful).

File: harness/determined/common/experimental/experiment.py [hunk 5]
```diff
@@ -115,6 +119,49 @@ def reload(self) -> None:
...
+            label: a string label to add to the experiment. If the label already exists,
+                the method call will be a no-op.
+        """
+
+        # URL-encode label for request.
+        label = parse.quote(label)
+        resp = bindings.put_PutExperimentLabel(
+            session=self._session, experimentId=self.id, label=label
+        )
+        self.labels = list(resp.labels)
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/common/experimental/experiment.py:135]: Which opinion are you looking for?
- I think that the server is just fundamentally broken because `set` mechanics cannot be safely used to implement `add/delete` mechanics, but add/delete mechanics are the only mechanics that we want to offer to end users.  So just fix the server.
- If the server were fixed, who cares if you replace the list or append to the list?  Maybe it matters because it's ordered?  Maybe a set would make more sense here, since that's really the experience we are trying to offer.
- In a world where the server isn't fixed, replacing self.labels with the freshest known value seems best, even if add_label() caused unexpected results to the ExperimentReference.labels value.  I'd rather be shown how other clients were breaking my logic then to be told things worked as I expected when really they didn't.

[the remark this answers is not in the document]
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:135]: Thanks, Ryan!

We've got PATCHes, PUTs, and POSTs all over our API with "set" semantics that serve as our standard way of adding/removing from collections. Until this comment, I don't think it'd occurred to me that this could be fixed. Even though that's exactly what Anda did right here.

I get a little confused with the "fix" language re: the server. At the moment, in this PR both `add_label` and `remove_label` both use add/remove methods on the backend. Even without "set", the local client's version can fall out of sync with the source of truth. I haven't figured out relevance of the antecedent in "In a world where the server isn't fixed, replacing self.labels with the freshest known value seems best."

Which is the precise question here I meant to pose:
1. Refresh the whole local snapshot of the collection?
2. Just add/remove the requested item?
3. Something else entirely?

I think you're right in weighing 1 vs. 2 to say that the user gets unhappy and surprised either way, and better to make that visible. I guess if we wanted we could also warn when the response indicates there's been a data update this client didn't know about.

File: master/internal/api_experiment.go (hunk 1 above)
Comment (azhou-determined) [reply master/internal/api_experiment.go]: fixed

File: harness/determined/common/experimental/experiment.py [hunk 6]
```diff
@@ -115,6 +119,49 @@ def reload(self) -> None:
...
+    def add_label(self, label: str) -> None:
+        """Add a label to the experiment.
+
+        Arguments:
+            label: a string label to add to the experiment. If the label already exists,
+                the method call will be a no-op.
+        """
+
+        # URL-encode label for request.
+        label = parse.quote(label)
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:131]: no, i don't think so, GRPC handles that. we only need to do this because it gets passed in the URL path.

File: harness/determined/common/experimental/experiment.py [hunk 7]
```diff
@@ -115,6 +119,49 @@ def reload(self) -> None:
...
+
+        This method will overwrite any existing labels on the experiment with the specified
+        labels.
+
+        Arguments:
+            labels: a list of string labels to set on the experiment. Duplicates will be ignored.
+        """
+        patch_exp = bindings.v1PatchExperiment(id=self.id, labels=labels)
+        resp = bindings.patch_PatchExperiment(self._session, body=patch_exp, experiment_id=self.id)
+        assert resp.experiment
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:162]: we need `Unset`, i think probably because we have no good way to tell the difference between `None` and "undefined" values, as in whether it was explicitly set to None, or not set at all.

File: harness/determined/common/experimental/experiment.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/common/experimental/experiment.py:131]: can we have a ticket to have our auto-generated grpc bindings.py be responsible for this?

File: harness/determined/common/experimental/experiment.py [hunk 8]
```diff
@@ -107,6 +110,7 @@ def id(self) -> int:
     def _hydrate(self, exp: bindings.v1Experiment) -> None:
         self.config = exp.config
         self.state = exp.state
+        self.labels = list(exp.labels) if exp.labels else None
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py]: yeah, i think this is a good idea. done.

File: master/internal/api_experiment.go [hunk 9]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+		Where("e.id = ?", req.ExperimentId).
+		Limit(1)
+
+	if err = query.Scan(ctx); err != nil {
+		return nil, err
+	}
+	modelExp, err := model.ExperimentFromProto(exp)
+	if err != nil {
+		return nil, err
+	}
```
Comment (stoksc) [inline master/internal/api_experiment.go:2374]: nit:
```suggestion
	exp := &experimentv1.Experiment{}
	query := db.Bun().NewSelect().
		ModelTableExpr("experiments as e").
		Model(exp).
		Apply(getExperimentColumns).
		Where("e.id = ?", req.ExperimentId).
		Limit(1)
	if err = query.Scan(ctx); err != nil {
		return nil, err
	}

	modelExp, err := model.ExperimentFromProto(exp)
	if err != nil {
		return nil, err
	}
```

File: master/internal/api_experiment.go [hunk 10]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+			logrus.WithError(err).Error("error rolling back transaction in create workspace")
+		}
+	}()
+
+	exp := &experimentv1.Experiment{}
+	query := db.Bun().NewSelect().
+		ModelTableExpr("experiments as e").
+		Model(exp).
+		Apply(getExperimentColumns).
+		Where("e.id = ?", req.ExperimentId).
```
Comment (stoksc) [inline master/internal/api_experiment.go:2365]: i think we need to `SELECT ... FOR UPDATE` here or we have the same problem on an even smaller level.

Comment (stoksc) [reply master/internal/api_experiment.go:2365]: at least that blocks other `SELECT ... FOR UPDATE`s on the same row, whereas if you to `begin transaction isolation level serializable` we have to retry or return 409 when we get a `could not serialize access due to concurrent update`.

Comment (stoksc) [reply master/internal/api_experiment.go:2365]: actually. i think `begin transaction isolation level serializable` and retries are probably the way to go. i think i prefer that over the issues `for update` can bring. curious what @ioga thinks too?

Comment (ioga) [reply master/internal/api_experiment.go:2365]: I believe `SELECT FOR UPDATE` and then `UPDATE` is sufficient. that's the pattern I've been using elsewhere.

- why would we need serializable transactions? if any other reader reads this row, they'll either see the old version, or the new one, which is fine. if another writer was updating it, they need to do `select for update` themselves, which will block until this transaction.
- not sure what would cause retries here.

Comment (stoksc) [reply master/internal/api_experiment.go:2365]: > why would we need serializable transactions?

you can do just `SELECT` without `FOR UPDATE` then `UPDATE` with `isolation level serializable` and then one of the transactions will get `could not serialize access due to concurrent update` and need to retry, instead of blocking with `FOR UPDATE`. I guess `FOR UPDATE` is probably a better API experience.

Comment (stoksc) [reply master/internal/api_experiment.go:2365]: But that way you can let pg be smart rather than devs with `FOR UPDATE`.

Comment (stoksc) [reply master/internal/api_experiment.go:2365]: (only one of the txs gets aborted due to deadlock, though, and the other is free to finish, so ultimately the approaches are kinda the same)

Comment (ioga) [reply master/internal/api_experiment.go:2365]: >I guess FOR UPDATE is probably a better API experience.

I agree, adding `.For("UPDATE")` is much simpler than having to wrap all transactions in retries.

>just because you can deadlock with FOR UPDATE

but that can't happen for this case with a single row update, correct?

Comment (stoksc) [reply master/internal/api_experiment.go:2365]: > but that can't happen for this case with a single row update, correct?

yeah, cannot.

File: harness/determined/common/experimental/experiment.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:131]: https://hpe-aiatscale.atlassian.net/browse/MLG-781

File: master/internal/api_experiment.go (hunk 10 above)
Comment (azhou-determined) [reply master/internal/api_experiment.go:2365]: added a FOR UPDATE.

File: master/internal/api_experiment.go [hunk 11]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+		For("UPDATE").
+		Limit(1)
```
Comment (stoksc) [inline master/internal/api_experiment.go:2427]: limit seems superfluous

File: master/internal/api_experiment.go [hunk 12]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+	exp.Labels = append(exp.Labels, req.Label)
+
+	_, err = tx.NewUpdate().Model(modelExp).
+		Set("config = jsonb_set(config, '{labels}', ?, true)", exp.Labels).
+		Where("id = ?", exp.Id).
+		Exec(ctx)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "error updating experiment %v in database", exp.Id)
+	}
```
Comment (stoksc) [inline master/internal/api_experiment.go:2393]:
```suggestion
	_, err = tx.NewUpdate().Model(modelExp).
		Set("config = jsonb_set(config, '{labels}', ?, true)", exp.Labels).
		Where("id = ?", exp.Id).
		Exec(ctx)
	if err != nil {
		return nil, errors.Wrapf(err, "error updating experiment %v in database", exp.Id)
	}
```

File: master/internal/api_experiment.go [hunk 13]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+
+	if err = query.Scan(ctx); err != nil {
+		return nil, err
+	}
+	modelExp, err := model.ExperimentFromProto(exp)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = exputil.AuthZProvider.Get().CanEditExperimentsMetadata(
```
Comment (stoksc) [inline master/internal/api_experiment.go:2376]: it's a shame this doesn't use the tx

File: master/internal/api_experiment.go (hunk 12 above)
Comment (stoksc) [inline master/internal/api_experiment.go:2393]:
```suggestion
	_, err = tx.NewUpdate().Model(modelExp).
		Set("config = jsonb_set(config, '{labels}', ?, true)", exp.Labels).
		Where("id = ?", exp.Id).
		Exec(ctx)
	if err != nil {
		return nil, errors.Wrapf(err, "error updating experiment %v in database", exp.Id)
	}
```

File: master/internal/api_experiment.go [hunk 14]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+		}
+	}()
+
+	exp := &experimentv1.Experiment{}
+	query := db.Bun().NewSelect().
+		ModelTableExpr("experiments as e").
+		Model(exp).
+		Apply(getExperimentColumns).
+		Where("e.id = ?", req.ExperimentId).
+		Limit(1)
```
Comment (stoksc) [inline master/internal/api_experiment.go:2366]: limit seems superfluous, does bun need this or something?

File: master/internal/api_experiment.go (hunk 9 above)
Comment (stoksc) [inline master/internal/api_experiment.go:2374]:
```suggestion
	query := db.Bun().NewSelect().
		ModelTableExpr("experiments as e").
		Model(exp).
		Apply(getExperimentColumns).
		Where("e.id = ?", req.ExperimentId).
		Limit(1)
	if err = query.Scan(ctx); err != nil {
		return nil, err
	}

	modelExp, err := model.ExperimentFromProto(exp)
	if err != nil {
		return nil, err
	}
```

File: master/internal/api_experiment.go [hunk 15]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+		return nil, status.Errorf(codes.PermissionDenied, err.Error())
+	}
+
+	for i, label := range exp.Labels {
+		if label == req.Label {
+			// Remove label from experiment.
+			exp.Labels = append(exp.Labels[:i], exp.Labels[i+1:]...)
+			break
+		}
+	}
```
Comment (stoksc) [inline master/internal/api_experiment.go:2448]:
```suggestion

	i := slices.Index(exp.Labels, req.Label)
	if i == -1 {
		return &apiv1.DeleteExperimentLabelResponse{Labels: exp.Labels}, nil
	}
	slices.Delete(xs, i, i+1)
```

File: master/internal/api_experiment.go [hunk 16]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+
+	_, err = tx.NewUpdate().Model(modelExp).
+		Set("config = jsonb_set(config, '{labels}', ?, true)", exp.Labels).
+		Where("id = ?", exp.Id).
+		Exec(ctx)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "error updating experiment %v in database", exp.Id)
+	}
```
Comment (stoksc) [inline master/internal/api_experiment.go:2457]:
```suggestion
	_, err = tx.NewUpdate().Model(modelExp).
		Set("config = jsonb_set(config, '{labels}', ?, true)", exp.Labels).
		Where("id = ?", exp.Id).
		Exec(ctx)
	if err != nil {
		return nil, errors.Wrapf(err, "error updating experiment %v in database", exp.Id)
	}
```

File: master/internal/api_experiment.go [hunk 17]
```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
...
+
+	if err = tx.Commit(); err != nil {
+		return nil, errors.Wrap(err, "could not commit delete experiment labels transaction")
```
Comment (stoksc) [inline master/internal/api_experiment.go:2460]: here and elsewhere
```suggestion
		return nil, fmt.Error("could not commit delete experiment labels transaction: %w", err)
```

**Review by stoksc — APPROVED** [review]
go side lgtm, just a bunch of style nits

File: master/internal/api_experiment.go (hunk 14 above)
Comment (azhou-determined) [reply master/internal/api_experiment.go]: nah, i thought it made sense as an extra precaution, but it doesn't, since the columns are unique by id anyway. removed.

File: master/internal/api_experiment.go [hunk 18]
```diff
@@ -2337,3 +2338,121 @@ func (a *apiServer) CreateTrial(
...
+	if i == -1 {
+		return &apiv1.DeleteExperimentLabelResponse{Labels: exp.Labels}, nil
+	}
+	exp.Labels = slices.Delete(exp.Labels, i, i+1)
+
+	_, err = tx.NewUpdate().Model(modelExp).
+		Set("config = jsonb_set(config, '{labels}', ?, true)", exp.Labels).
+		Where("id = ?", exp.Id).
+		Exec(ctx)
+
```
Comment (stoksc) [inline master/internal/api_experiment.go:2448]:
```suggestion
```

**Review by stoksc — APPROVED** [review]

File: harness/tests/determined/common/experimental/test_experiment.py [hunk 19]
```diff
@@ -209,3 +210,38 @@ def test_unarchive_doesnt_update_local_on_rest_failure(
...
+    assert parse.unquote(url_encoded_label) == label
+
+    expref.add_label(label=label)
+
+    assert mock_bindings.call_count == 1
+    _, call_kwargs = mock_bindings.call_args_list[0]
+    assert call_kwargs["label"] == url_encoded_label
+
+
+@mock.patch("determined.common.api.bindings.delete_DeleteExperimentLabel")
```
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:233]: I don't remember what you decided, re: adding tests that labels don't get modified if there's a network failure. We talked about it... right? I'm good either way, just confused that I don't see a reference to it in this CR.

File: harness/tests/determined/common/experimental/test_experiment.py [hunk 20]
```diff
@@ -209,3 +210,38 @@ def test_unarchive_doesnt_update_local_on_rest_failure(
...
+    make_expref: Callable[[int], experiment.Experiment],
+) -> None:
+    expref = make_expref(1)
+    label = "label with @#$%^&* symbols"
+    url_encoded_label = "label%20with%20%40%23%24%25%5E%26%2A%20symbols"
+    assert parse.unquote(url_encoded_label) == label
+
+    expref.add_label(label=label)
+
+    assert mock_bindings.call_count == 1
```
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:228]: I dunno... I get that this is how the code should work, but the assertion just doesn't seem relevant to the test. My own opinion here is that if you think "one call to bindings" is worth testing, then I think it should be its own test.

File: harness/determined/common/experimental/experiment.py [hunk 21]
```diff
@@ -148,6 +154,49 @@ def set_notes(self, notes: str) -> None:
...
+                exist on the experiment, this method call will be a no-op.
+        """
+        # URL-encode label for request.
+        label = parse.quote(label)
+        resp = bindings.delete_DeleteExperimentLabel(
+            session=self._session, experimentId=self.id, label=label
+        )
+        self.labels = set(resp.labels)
+
+    def set_labels(self, labels: Set[str]) -> None:
```
[the remark this answers is not in the document]
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:186]: Also, FWIW, if you find it inconvenient you're still compliant with the style guide to omit the type altogether:

> The description should include required type(s) if the code does not contain a corresponding type annotation

The prescription here is more evolved than set, coming back from before Python got typing.

**Review by wes-turner — APPROVED** [review]

File: harness/tests/determined/common/experimental/test_experiment.py (hunk 4 above)
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_experiment.py:215]: 🤦

File: harness/determined/common/experimental/experiment.py [hunk 22]
```diff
@@ -104,6 +106,8 @@ def __init__(
         # These properties may be mutable and will be set by _hydrate()
         self.config: Optional[Dict[str, Any]] = None
         self.state: Optional[bindings.experimentv1State] = None
+        self.labels: Optional[Set[str]] = None
+        self.labels: Optional[List[str]] = None
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:110]: 🤦 this was due to a tricky merge rebase, thanks for catching this.

File: harness/determined/common/experimental/experiment.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:186]: i'm not sure why the bindings seem to prefer Sequence, but i think Set is the sensible choice here. i updated the docstrings.

File: harness/tests/determined/common/experimental/test_experiment.py (hunk 19 above)
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_experiment.py:232]: unless i'm misremembering, we talked about that, but for a different feature (adding archive/unarchive on experiment), and those tests are right above this one (test_unarchive_doesnt_update_local_on_rest_failure)

File: harness/tests/determined/common/experimental/test_experiment.py (hunk 20 above)
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_experiment.py]: i think you're right. i added them more out of anxiety ("but what if?") than practicality. 😄  removed
